### ISSUE TITLE
Shared calendars / contacts and fine grained permissions

### ISSUE THREAD

**Issue by evert** [fruux/Baikal2#30]
_From @netgusto on November 26, 2014 8:19_

B2 should really support these !

_Copied from original issue: fruux/Baikal#303_

**Comment by evert** [comment]
_From @schmidp on January 20, 2015 8:48_

is this something the development version already supports?

**Comment by evert** [comment]
_From @mrbaseman on March 7, 2015 20:7_

for managing the permissions to calendars across user accounts in Baikal 0.2.7 I have written a script: https://github.com/mrbaseman/calendar-tools/blob/master/permissions.php (just leaving this comment here for others who are searching for such a tool and stumble across this ticket)

**Comment by evert** [comment]
_From @SamuelMoraesF on March 30, 2015 20:5_

Would be fantastic an option to share calendars with public link(read-only mode, without login)

**Comment by evert** [comment]
_From @mrbaseman on March 31, 2015 19:57_

actually, this is a special case included in my script. Just add a user, let's call him "web" and assign a simple password to him (e.g. "readonly" - it doesn't matter at all) and remove the default calendar and carddav ressource. So, the user web has no writable ressources at all. Then, use my script to grant him read-access to other user's calendars. The link to access the default calendar of user called "user" is: http://web:readonly@www.examle.com/baikal/cal.php/calendars/user/default

**Comment by evert** [comment]
_From @dakira on May 21, 2015 8:29_

@mrbaseman, thanks for this! Exactly what I was looking for! Just a hint through: You should start to get used to PDO and use prepared statements to prevent SQL injections.

**Comment by evert** [comment]
_From @mrbaseman on June 3, 2015 5:21_

@dakira,
> Just a hint through: You should start to get used to PDO and use prepared statements to prevent SQL injections.
Thanks for the hint. I have been thinking about it for a while, but I
don't quite understand how SQL injections would be an attack vector in
this case. The first thing I do (after including some definitions), is
to check if the user is properly authenticated as admin and I check the
password hash. Without knowledge of that, the script exits in the first
place. Note that the credentials are not stored in the database. They
are defined in constants to which I compare the input coming from the
web server. Only, if this is correctly verified, a connection to the
database is opened.
Then, if you are already authenticated as admin, you anyway have full
access and there is no need to protect against SQL injections. Note:
there is no code sent to the browser for local execution (Javascript and
the like), so also no danger from XSS.

**Comment by evert** [comment]
_From @dakira on June 4, 2015 14:15_

@mrbaseman it's just a weak link that's avoidable. If you look at the recent Magento attack you'll see that every individual component was secure and well thought out. There were several issues that couldn't be used for an attack individually. Only chained together, an attack was possible. Not using prepared statements could be part of such a chain (even if you can't see an attack yet).

**Comment by evert** [comment]
_From @zot24 on June 26, 2015 16:7_

How @netgusto are you going to implement the sharing addressbooks between users for a CardDav server? 

I was just having a look to your code on [BaikalModelBundle](https://github.com/netgusto/BaikalModelBundle) and I was wondering if you are gonna use [GroupMembers](https://github.com/netgusto/BaikalModelBundle/blob/master/Entity/Groupmember.php) as a kind of pivot table to handle the relations between the principals (users) and the principals (groups of users) to achieve the sharing of an addressbook between the members of a group.

btw great project thanks

**Comment by evert** [comment]
_From @m0yellow on January 7, 2016 10:22_

@mrbaseman there is no excuse for not using prepared statements, also in the admin menu.
To illustrate that think the following example: even with XSRF, cookie leak or man-in-the-browser attack you could still implement read/export limit, strong history, query log, and the attack surface would be much smaller. If I can hijack the admin with a simple phishing link, I could download and delete the whole calendar in most companies, as there tend to be more admins than needed for install (e.g. HR, for assigning rights).

TL;DR: +1 for proper permission handling in Baikal, as this is the backbone for professional use-cases.

**Comment by evert** [comment]
_From @mrbaseman on January 9, 2016 20:8_

@Matthias first of all, note that I only have provided the user
management script as a workaround. I haven't been working on other
things concerning the admin menu, yet (because I didn't have the time to
learn how to use the Flake class framework.

Coming back to your main point about prepared statements in the
permission administration script: prepared statements help against first
order sql injections when user supplied data is used in queries. But
that's not the case here. I have queries without user-supplied
components to build up the lists of users, groups, and permissions. Only
these IDs are compared to the posted values. For setting the permissions
I also don't use the posted variables directly in the queries. I only
use them to decide which 'if'-branch to go. The statement itself again
is based on values originally coming directly out of the database. So,
prepared statements would not improve anything here, because the attacks
prevented by prepared statements wouldn't be successful anyway. Prepared
statements also wouldn't help against a man in the browser attack.
Thee is already some protection in place against taking over the admin
session, because the hash value contains the client IP and it expires
every midnight.

Anyhow, I can rewrite the script with prepared statements. It seems that
this is a strong desire for some users...

**Comment by evert** [comment]
_From @m0yellow on January 9, 2016 21:41_

thanks for your lengthy and elaborated reply,
as I came here via google, I wanted to give advice more for the anonymous reader.

Please take it as a reference to best current practice. 
I appreciate your work, and my ideas about probing and take overs are only from the daily reading at my abuse@<provider> address.
